android/canonical/app/src/main/java/com/google/samples/quickstart/canonical/LoginFragment.kt

```diff
@@ -0,0 +1,130 @@
+package com.google.samples.quickstart.canonical
+
+import android.content.Intent
+import android.os.Bundle
+import android.util.Log
+import androidx.fragment.app.Fragment
+import android.view.LayoutInflater
+import android.view.View
+import android.view.ViewGroup
+import androidx.lifecycle.ViewModelProviders
+import com.google.android.gms.auth.api.signin.GoogleSignIn
+import com.google.android.gms.common.SignInButton
+import com.google.android.gms.common.api.ApiException
+import com.google.firebase.auth.GoogleAuthProvider
+
+// TODO: Rename parameter arguments, choose names that match
+// the fragment initialization parameters, e.g. ARG_ITEM_NUMBER
+private const val ARG_PARAM1 = "param1"
+private const val ARG_PARAM2 = "param2"
+
+/**
+ * A simple [Fragment] subclass.
+ * Use the [LoginFragment.newInstance] factory method to
+ * create an instance of this fragment.
+ */
+class LoginFragment : Fragment() {
+    // TODO: Rename and change types of parameters
+    private var param1: String? = null
+    private var param2: String? = null
+    private lateinit var signInVM : SignInViewModel
+
+    private fun signIn() {
+        val signInIntent = signInVM.getGoogleSignInClient().signInIntent
+        startActivityForResult(signInIntent, RC_SIGN_IN)
+    }
+
+    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
+        super.onActivityResult(requestCode, resultCode, data)
+
+        // Result returned from launching the Intent from GoogleSignInApi.getSignInIntent(...);
+        if (requestCode == RC_SIGN_IN) {
+            val task = GoogleSignIn.getSignedInAccountFromIntent(data)
+            if (task.isSuccessful) {
+                try {
+                    // Google Sign In was successful, authenticate with Firebase
+                    val account = task.getResult(ApiException::class.java)!!
+                    Log.d(TAG, "Google Sign In was successful:" + account.id)
+                    firebaseAuthWithGoogle(account.idToken!!)
+                    val intent = Intent(context, MainActivity::class.java)
+                    intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP)
+                    startActivity(intent)
+                } catch (e: ApiException) {
+                    // Google Sign In failed
+                    Log.w(TAG, "Google sign in failed", e)
+                }
+            } else {
+                Log.w(TAG, "Google sign in unsuccessful")
+            }
```

Comment:
I don't want to speak for Arco, but I think she meant all of this logic including firebaseAuthWithGoogle could be refactored so that the Fragment doesn't know about GoogleSignIn.getSignedInAccountFromIntent or other sign-in logic in firebaseAuthWithGoogle.

```
if (requestCode == RC_SIGN_IN) {
  signInVM.handleSignInActivityResult(data);
}
```

Then in the Fragment you could observe live data https://developer.android.com/topic/libraries/architecture/viewmodel#implement

That live data could be a data class that contains the account plus a status enum or other indication of sign-in status e.g. SIGNED_IN/SIGNED_OUT. https://kotlinlang.org/docs/reference/data-classes.html#data-classes